File: questPresentation/src/main/java/ru/nekit/android/qls/window/common/QuestWindow.java
package ru.nekit.android.qls.window.common;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StyleRes;

import ru.nekit.android.qls.quest.QuestContext;
import ru.nekit.android.qls.window.Window;
import ru.nekit.android.qls.window.WindowContentViewHolder;
import ru.nekit.android.qls.window.WindowListener;

public class QuestWindow extends Window {

    public static final String VALUE_WINDOW_NAME = "ru.nekit.android.qls.value_window_name";
    public static final String EVENT_WINDOW_OPEN = "ru.nekit.android.qls.event_window_open";
    public static final String EVENT_WINDOW_OPENED = "ru.nekit.android.qls.event_window_opened";
    public static final String EVENT_WINDOW_CLOSE = "ru.nekit.android.qls.event_window_close";
    public static final String EVENT_WINDOW_CLOSED = "ru.nekit.android.qls.event_window_closed";

    private static WindowListener mWindowListener = new WindowListener() {

        @Override
        public void onWindowOpen(@NonNull Window window) {
            sendEvent(window.getName(), EVENT_WINDOW_OPEN);
        }

        @Override
        public void onWindowOpened(@NonNull Window window) {
            sendEvent(window.getName(), EVENT_WINDOW_OPENED);
        }

        @Override
        public void onWindowClose(@NonNull Window window) {
            sendEvent(window.getName(), EVENT_WINDOW_CLOSE);
        }

        @Override
        public void onWindowClosed(@NonNull Window window) {
            sendEvent(window.getName(), EVENT_WINDOW_CLOSED);
        }
    };

    public QuestWindow(@NonNull QuestContext context,
                       @Nullable String name,
                       @Nullable WindowContentViewHolder content,
                       @StyleRes int styleResId) {
        super(context, name, mWindowListener, content, styleResId);
    }

    private static void sendEvent(@NonNull String windowName, @NonNull String eventName) {
        QuestContext.getInstance().getEventBus().sendEvent(eventName, VALUE_WINDOW_NAME, windowName);
    }

}

File: questPresentation/src/main/java/ru/nekit/android/qls/window/common/QuestWindow.kt
package ru.nekit.android.qls.window.common

import android.support.annotation.StyleRes
import ru.nekit.android.domain.event.IEvent
import ru.nekit.android.domain.event.IEventSender
import ru.nekit.android.qls.quest.QuestContext
import ru.nekit.android.qls.window.common.QuestWindowEvent.*
import ru.nekit.android.window.Window
import ru.nekit.android.window.WindowContentViewHolder
import ru.nekit.android.window.WindowListener

open class QuestWindow(questContext: QuestContext,
                       name: String,
                       content: WindowContentViewHolder?,
                       @StyleRes styleResId: Int) : Window(questContext, name, object : WindowListener {

    override fun onWindowOpen(window: Window) {
        sendEvent(questContext.eventSender, OPEN, window.name)
    }

    override fun onWindowOpened(window: Window) {
        sendEvent(questContext.eventSender, OPENED, window.name)
    }

    override fun onWindowClose(window: Window) {
        sendEvent(questContext.eventSender, CLOSE, window.name)
    }

    override fun onWindowClosed(window: Window) {
        sendEvent(questContext.eventSender, CLOSED, window.name)
    }
}, content, styleResId) {

    companion object {
        private fun sendEvent(eventSender: IEventSender, event: QuestWindowEvent, windowName: String) {
            event.windowName = windowName
            eventSender.send(event)
        }
    }
}

enum class QuestWindowEvent : IEvent {

    OPEN,
    OPENED,
    CLOSE,
    CLOSED;

    lateinit var windowName: String

    override val eventName: String = "${javaClass.name}::$name"

}